packages/angular/ssr/src/i18n.ts

```diff
@@ -43,3 +43,147 @@ export function getPotentialLocaleIdFromUrl(url: URL, basePath: string): string
   // Extract the potential locale id.
   return pathname.slice(start, end);
 }
+
+/**
+ * Parses the `Accept-Language` header and returns a list of locale preferences with their respective quality values.
+ *
+ * The `Accept-Language` header is typically a comma-separated list of locales, with optional quality values
+ * in the form of `q=<value>`. If no quality value is specified, a default quality of `1` is assumed.
+ * Special case: if the header is `*`, it returns the default locale with a quality of `1`.
+ *
+ * @param header - The value of the `Accept-Language` header, typically a comma-separated list of locales
+ *                  with optional quality values (e.g., `en-US;q=0.8,fr-FR;q=0.9`). If the header is `*`,
+ *                  it represents a wildcard for any language, returning the default locale.
+ *
+ * @returns A `ReadonlyMap` where the key is the locale (e.g., `en-US`, `fr-FR`), and the value is
+ *          the associated quality value (a number between 0 and 1). If no quality value is provided,
+ *          a default of `1` is used.
+ *
+ * @example
+ * ```js
+ * parseLanguageHeader('en-US;q=0.8,fr-FR;q=0.9')
+ * // returns new Map([['en-US', 0.8], ['fr-FR', 0.9]])
+
+ * parseLanguageHeader('*')
+ * // returns new Map([['*', 1]])
+ * ```
+ */
+function parseLanguageHeader(header: string): ReadonlyMap<string, number> {
+  if (header === '*') {
+    return new Map([['*', 1]]);
+  }
+
+  const parsedValues = header
+    .split(',')
+    .map((item) => {
+      const [locale, qualityValue] = item
+        .trim()
+        .split(';', 2)
+        .map((v) => v.trim());
+
+      let quality = qualityValue?.startsWith('q=') ? parseFloat(qualityValue.slice(2)) : undefined;
+      if (typeof quality !== 'number' || isNaN(quality) || quality < 0 || quality > 1) {
+        quality = 1; // Invalid quality value defaults to 1
+      }
+
+      return [locale, quality] as const;
+    })
+    .sort((a, b) => b[1] - a[1]);
```

Comment:
Question: Is there any specced behavior for how to break ties? What should we do if someone sends `Accept-Language: en;q=0.9, it;q=0.9`? Do we take `en` because it's first? Is that an invalid request?

Reply:
When multiple languages in the Accept-Language header have the same quality value, the order of appearance determines priority. This behavior aligns with the HTTP specification and is widely supported across servers. According to RFC standards, the header specifies language preferences with optional quality values (q), and the server processes them as follows:

- Languages with the highest quality (q) values are prioritized.
- If multiple languages have equal q values, the order in the header determines which one is preferred.